## Условие 1:
### 1 задача
- Скачать файл в уроке
- Считать данные с помощью pandas
- Вывести на экран первые 5 строк
- Посмотреть на описание признаков и на их содержание

In [249]:
import scipy.stats as stats
import numpy as np
import matplotlib.pyplot as plt
from sklearn. linear_model import LinearRegression
from scipy.stats import bartlett
import platform
from typing import List, Dict
import pandas as pd
import datetime as dt


In [250]:
#df = pd.read_csv('kc_house_data (1).csv')
df = pd.read_csv('kc_house_data.csv', sep=',', encoding ='cp1251', quoting=3)

In [251]:
df.head(5)

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

## Условие 2:
### 2 задача

- Проведите первичный анализ данных
- Изучите типы данных
- Найдите количество пропущенных ячеек в данных
- Посчитайте основные статистики по всем признакам и поизучайте их
- Пишите выводы

In [252]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

**Вывод:** пропусков в данных таблицы нет. Типы данных в таблице 3: int64, float64, object. 

In [253]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

## Условие 3:
### 3 задача

3.1 В каком диапазоне изменяются стоимости недвижимости?

In [254]:
df['price'].min()

75000.0

**Вывод:** Минимальная стоимость 75000.

In [255]:
df['price'].max()

7700000.0

**Вывод:** Максимальная стоимость 7700000.

В диапазоне от 75000 до 77000000 изменяются стоимости недвижимости.

3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [256]:
p =2079.899736/1.510697e+04
p

0.1376781535940033

**Вывод:** В среднем жилая площадь занимает от всей площади по всем домам - 14%

3.3 Как много домов с разными этажами в данных?

In [257]:
df['floors'].unique()

array([1. , 2. , 1.5, 3. , 2.5, 3.5])

In [258]:
df['floors'].nunique()

6

**Вывод:** 6 значений количества этажей в данных: 1, 2, 1.5, 3, 2.5, 3.5.

3.4 Насколько хорошие состояния у домов в данных?

In [259]:
df['condition'].value_counts(normalize=True)

3    0.649193
4    0.262759
5    0.078703
2    0.007958
1    0.001388
Name: condition, dtype: float64

3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [260]:
df['yr_built'].min()

1900

In [261]:
df['yr_built'].max()

2015

**Вывод:** Первый дом построили в 1900 году, а последний в 2015 году.

## Условие 4:
### 4 задача

Ответьте на несколько вопросов

4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [262]:
df[df['bedrooms'] == 2]['price'].mean()

401372.681884058

**Вывод:** Дома, у которых 2 спальни в среднем стоят 401373.

4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [263]:
df[df['price'] > 600000]['sqft_lot'].mean()

20442.524776214832

**Вывод:** В среднем общая площадь домов, у которых стоимость больше 600000 равна 20443.

4.3 Как много домов коснулся ремонт?

In [264]:
df[df['yr_renovated']> 0]

id             date      price  bedrooms  bathrooms  \
1      6414100192  20141209T000000   538000.0         3       2.25   
35     9547205180  20140613T000000   696000.0         3       2.50   
95     1483300570  20140908T000000   905000.0         4       2.50   
103    2450000295  20141007T000000  1088000.0         3       2.50   
115    3626039325  20141121T000000   740500.0         3       3.50   
...           ...              ...        ...       ...        ...   
19622  7351200295  20150114T000000  1150000.0         3       1.75   
20057   126039256  20140904T000000   434900.0         3       2.00   
20444  4305600360  20150225T000000   500012.0         4       2.50   
20447  3319500628  20150212T000000   356999.0         3       1.50   
20962  1278000210  20150311T000000   110000.0         2       1.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
1             2570      7242     2.0           0     0  ...      7   
35            2300      3060     1.5           0     0  ...      8   
95            3300     10250     1.0           0     0  ...      7   
103           2920      8113     2.0           0     0  ...      8   
115           4380      6350     2.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
19622         1760      6788     2.0           1     4  ...      7   
20057         1520      5040     2.0           0     0  ...      7   
20444         2400      9612     1.0           0     0  ...      8   
20447         1010      1546     2.0           0     0  ...      8   
20962          828      4524     1.0           0     0  ...      6   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
1            2170            400      1951          1991    98125  47.7210   
35           1510            790      1930          2002    98115  47.6827   
95           2390            910      1946          1991    98040  47.5873   
103          2920              0      1950          2010    98004  47.5814   
115          2780           1600      1900          1999    98117  47.6981   
...           ...            ...       ...           ...      ...      ...   
19622        1760              0      1940          1960    98125  47.7336   
20057        1520              0      1977          2006    98177  47.7770   
20444        1230           1170      1962          2009    98059  47.4799   
20447        1010              0      1971          2014    98144  47.5998   
20962         828              0      1968          2007    98001  47.2655   

          long  sqft_living15  sqft_lot15  
1     -122.319           1690        7639  
35    -122.310           1590        3264  
95    -122.249           1950        6045  
103   -122.196           2370        8113  
115   -122.368           1830        6350  
...        ...            ...         ...  
19622 -122.284           1630        7588  
20057 -122.362           1860        8710  
20444 -122.127           2430        5539  
20447 -122.311           1010        1517  
20962 -122.244            828        5402  

[914 rows x 21 columns]

**Вывод:** 914 домов отремонтированы.

4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [265]:
df[df['grade'] > 10]['price'].mean()-df[df['grade'] < 4]['price'].mean()


1488885.1175298805

**Вывод:** На 1488885 в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4.

## Условие 5: 
### 5 задача

Ответьте на несколько вопросов

5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?

In [266]:
df[(df['view'] == 'waterfront') & (df['bathrooms'] == 3) & (df['sqft_basement'] == 1)]

Empty DataFrame
Columns: [id, date, price, bedrooms, bathrooms, sqft_living, sqft_lot, floors, waterfront, view, condition, grade, sqft_above, sqft_basement, yr_built, yr_renovated, zipcode, lat, long, sqft_living15, sqft_lot15]
Index: []

[0 rows x 21 columns]

**Вывод:** Подходящего дома в данных нет.

5.2 Выберите дом клиенту

Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [320]:
df1 = df[(df['view'] == 'waterfront') | (df['view'] == 4) & (df['yr_built'] > 1980)]
df1

id             date      price  bedrooms  bathrooms  \
246    2025069065  20140929T000000  2400000.0         4       2.50   
300    3225069065  20140624T000000  3075000.0         4       5.00   
409    1560930070  20140911T000000   840000.0         4       3.50   
757    9273200140  20150121T000000  1310000.0         2       2.25   
1164   1247600105  20141020T000000  5110800.0         5       5.25   
...           ...              ...        ...       ...        ...   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
20923  7852070210  20140527T000000  1149000.0         4       3.00   
21050  2424059170  20150219T000000   900000.0         5       6.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
246           3650      8354     1.0           1     4  ...      9   
300           4550     18641     1.0           1     4  ...     10   
409           2840     40139     1.0           0     4  ...     10   
757           3950      3938     2.0           0     4  ...     10   
1164          8010     45517     2.0           1     4  ...     12   
...            ...       ...     ...         ...   ...  ...    ...   
20767         4360      8175     2.5           1     4  ...     10   
20923         5940     11533     2.0           0     4  ...     11   
21050         7120     40806     2.0           0     4  ...     12   
21201         3760      5634     2.0           1     4  ...     11   
21576         4850     10584     2.0           1     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
246          1830           1820      2000             0    98074  47.6338   
300          2600           1950      2002             0    98074  47.6053   
409          2840              0      1986             0    98038  47.4010   
757          2910           1040      1991             0    98116  47.5912   
1164         5990           2020      1999             0    98033  47.6767   
...           ...            ...       ...           ...      ...      ...   
20767        3940            420      2007             0    98008  47.5724   
20923        4950            990      2004             0    98065  47.5443   
21050        5480           1640      2007             0    98006  47.5451   
21201        2830            930      2014             0    98056  47.5285   
21576        3540           1310      2007             0    98008  47.5943   

          long  sqft_living15  sqft_lot15  
246   -122.072           3120       18841  
300   -122.077           4550       19508  
409   -122.026           3180       36852  
757   -122.384           3220        4500  
1164  -122.211           3430       26788  
...        ...            ...         ...  
20767 -122.104           2670        8525  
20923 -121.870           4240       12813  
21050 -122.114           3440       36859  
21201 -122.205           3560        5762  
21576 -122.110           3470       18270  

[108 rows x 21 columns]

In [321]:
df1['price'].min()

380000.0

In [322]:
df1['price'].max()

6885000.0

**Вывод:** Подходящих домов в данных 108, с диапазоном цен от 380000 до 6885000.

5.3 Выберите дом клиенту

Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [324]:
df2 = df[(df['sqft_basement'] == 0) & (df['floors'] == 2) & (df['price'] < 150000)]
df2

id             date     price  bedrooms  bathrooms  \
2589   5061300030  20140508T000000  134000.0         2       1.50   
10105  5466310060  20150324T000000  139500.0         2       1.50   
15610  7614100080  20150211T000000  140000.0         3       1.75   
18332  9272202260  20140924T000000  130000.0         3       1.00   
18939  4239400300  20141129T000000   90000.0         3       1.00   
18988  7399100210  20141126T000000  140000.0         3       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2589           980      5000     2.0           0     0  ...      7   
10105         1230      1561     2.0           0     0  ...      7   
15610         1270      8991     2.0           0     0  ...      7   
18332         1200      7000     2.0           0     0  ...      7   
18939          980      2490     2.0           0     0  ...      6   
18988         1200      2002     2.0           0     0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2589          980              0      1922          2003    98014  47.7076   
10105        1230              0      1983             0    98042  47.3768   
15610        1270              0      1981             0    98042  47.3563   
18332        1200              0      1908             0    98116  47.5883   
18939         980              0      1969             0    98092  47.3170   
18988        1200              0      1966             0    98055  47.4659   

          long  sqft_living15  sqft_lot15  
2589  -121.359           1040        5000  
10105 -122.149           1660        2243  
15610 -122.149           1270        8993  
18332 -122.384           3290        6000  
18939 -122.182            980        3154  
18988 -122.189           1270        1848  

[6 rows x 21 columns]

In [325]:
df2['condition'].mean()

2.8333333333333335

**Вывод:** Подходящих домов в данных 6, со средней оценкой по состоянию 2,83.